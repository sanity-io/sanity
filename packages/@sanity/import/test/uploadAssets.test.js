const path = require('path')
const fileUrl = require('file-url')
const noop = require('lodash/noop')
const nock = require('nock')
const uploadAssets = require('../src/uploadAssets')
const mockAssets = require('./fixtures/mock-assets')
const {getSanityClient} = require('./helpers')

const fixturesDir = path.join(__dirname, 'fixtures')
const imgFileUrl = fileUrl(path.join(fixturesDir, 'img.gif'))
const fileAsset = {
  documentId: 'movie_1',
  path: 'metadata.poster',
  type: 'image',
  url: imgFileUrl,
}

const fetchFailClient = {
  fetch: () => Promise.reject(new Error('Some network err')),
}

test('fails if asset download fails', () => {
  expect.assertions(1)
  const asset = Object.assign({}, fileAsset, {
    url: 'http://127.0.0.1:49999/img.gif',
  })

  return expect(uploadAssets([asset], {client: null, onProgress: noop})).rejects.toMatchSnapshot()
})

test('fails if asset lookup fails', async () => {
  const options = {client: fetchFailClient, onProgress: noop}
  try {
    const result = await uploadAssets([fileAsset], options)
    expect(result).toBeFalsy()
  } catch (err) {
    expect(err.message).toMatch(/Some network err/)
  }
})

test('will reuse an existing asset if it exists', () => {
  nock('https://foo.bar.baz').head('/images/foo/bar/someAssetId-200x200.png').reply(200)

  const client = getSanityClient((req) => {
    const options = req.context.options
    const uri = options.uri || options.url

    if (uri.includes('/data/query')) {
      return {
        body: {
          result: {
            _id: 'image-someAssetId',
            url: 'https://foo.bar.baz/images/foo/bar/someAssetId-200x200.png',
          },
        },
      }
    }

    if (uri.includes('/data/mutate')) {
      const body = JSON.parse(options.body)
      expect(body).toMatchSnapshot('single asset mutation')
      const results = body.mutations.map((mut) => ({
        id: mut.patch.id,
        operation: 'update',
      }))
      return {body: {results}}
    }

    return {statusCode: 400, body: {error: `"${uri}" should not be called`}}
  })

  return expect(uploadAssets([fileAsset], {client, onProgress: noop})).resolves.toMatchObject({
    batches: 1,
    failures: [],
  })
})

test('will upload an asset if asset doc exists but file does not', () => {
  nock('https://foo.bar.baz').head('/images/foo/bar/someAssetId-200x200.png').reply(404)

  const client = getSanityClient((req) => {
    const options = req.context.options
    const uri = options.uri || options.url

    if (uri.includes('/data/query')) {
      return {
        body: {
          result: {
            _id: 'image-someAssetId',
            url: 'https://foo.bar.baz/images/foo/bar/someAssetId-200x200.png',
          },
        },
      }
    }

    if (uri.includes('/assets/images')) {
      return {body: {document: {_id: 'image-newAssetId'}}}
    }

    if (uri.includes('/data/mutate')) {
      const body = JSON.parse(options.body)
      expect(body).toMatchSnapshot('single create mutation')
      const results = body.mutations.map((mut) => ({
        id: mut.patch.id,
        operation: 'update',
      }))
      return {body: {results}}
    }

    return {statusCode: 400, body: {error: `"${uri}" should not be called`}}
  })

  return expect(uploadAssets([fileAsset], {client, onProgress: noop})).resolves.toMatchObject({
    batches: 1,
    failures: [],
  })
})

test('will upload asset that do not already exist', () => {
  const client = getSanityClient((req) => {
    const options = req.context.options
    const uri = options.uri || options.url
    if (uri.includes('/data/query')) {
      return {body: {result: null}}
    }

    if (uri.includes('/assets/images')) {
      return {body: {document: {_id: 'image-newAssetId'}}}
    }

    if (uri.includes('/data/mutate')) {
      const body = JSON.parse(options.body)
      expect(body).toMatchSnapshot('single create mutation')
      const results = body.mutations.map((mut) => ({
        id: mut.patch.id,
        operation: 'update',
      }))
      return {body: {results}}
    }

    return {statusCode: 400, body: {error: `"${uri}" should not be called`}}
  })

  return expect(uploadAssets([fileAsset], {client, onProgress: noop})).resolves.toMatchObject({
    batches: 1,
    failures: [],
  })
})

test('will upload once but batch patches', () => {
  nock('https://foo.bar.baz').head('/images/foo/bar/someAssetId-200x200.png').reply(200)

  let batch = 0
  const client = getSanityClient((req) => {
    const options = req.context.options
    const uri = options.uri || options.url

    if (uri.includes('/data/query')) {
      return {
        body: {
          result: {
            _id: 'image-someAssetId',
            url: 'https://foo.bar.baz/images/foo/bar/someAssetId-200x200.png',
          },
        },
      }
    }

    if (uri.includes('/data/mutate')) {
      const body = JSON.parse(options.body)
      expect(body).toMatchSnapshot(`batch patching (batch #${++batch})`)
      const results = body.mutations.map((mut) => ({
        id: mut.patch.id,
        operation: 'update',
      }))
      return {body: {results}}
    }

    return {statusCode: 400, body: {error: `"${uri}" should not be called`}}
  })

  const upload = uploadAssets(mockAssets(imgFileUrl), {
    client,
    onProgress: noop,
  })
  return expect(upload).resolves.toMatchObject({
    batches: 60,
    failures: [],
  })
})
