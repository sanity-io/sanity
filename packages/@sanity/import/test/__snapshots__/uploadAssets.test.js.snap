// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`batch patching (batch #1) 1`] = `
Object {
  "mutations": Array [
    Object {
      "patch": Object {
        "id": "doc_1",
        "set": Object {
          "some.path": Object {
            "_ref": "someAssetId",
            "_type": "reference",
          },
        },
      },
    },
    Object {
      "patch": Object {
        "id": "doc_2",
        "set": Object {
          "some.path": Object {
            "_ref": "someAssetId",
            "_type": "reference",
          },
        },
      },
    },
    Object {
      "patch": Object {
        "id": "doc_3",
        "set": Object {
          "some.path": Object {
            "_ref": "someAssetId",
            "_type": "reference",
          },
        },
      },
    },
    Object {
      "patch": Object {
        "id": "doc_4",
        "set": Object {
          "some.path": Object {
            "_ref": "someAssetId",
            "_type": "reference",
          },
        },
      },
    },
    Object {
      "patch": Object {
        "id": "doc_5",
        "set": Object {
          "some.path": Object {
            "_ref": "someAssetId",
            "_type": "reference",
          },
        },
      },
    },
    Object {
      "patch": Object {
        "id": "doc_6",
        "set": Object {
          "some.path": Object {
            "_ref": "someAssetId",
            "_type": "reference",
          },
        },
      },
    },
    Object {
      "patch": Object {
        "id": "doc_7",
        "set": Object {
          "some.path": Object {
            "_ref": "someAssetId",
            "_type": "reference",
          },
        },
      },
    },
    Object {
      "patch": Object {
        "id": "doc_8",
        "set": Object {
          "some.path": Object {
            "_ref": "someAssetId",
            "_type": "reference",
          },
        },
      },
    },
    Object {
      "patch": Object {
        "id": "doc_9",
        "set": Object {
          "some.path": Object {
            "_ref": "someAssetId",
            "_type": "reference",
          },
        },
      },
    },
    Object {
      "patch": Object {
        "id": "doc_10",
        "set": Object {
          "some.path": Object {
            "_ref": "someAssetId",
            "_type": "reference",
          },
        },
      },
    },
    Object {
      "patch": Object {
        "id": "doc_11",
        "set": Object {
          "some.path": Object {
            "_ref": "someAssetId",
            "_type": "reference",
          },
        },
      },
    },
    Object {
      "patch": Object {
        "id": "doc_12",
        "set": Object {
          "some.path": Object {
            "_ref": "someAssetId",
            "_type": "reference",
          },
        },
      },
    },
    Object {
      "patch": Object {
        "id": "doc_13",
        "set": Object {
          "some.path": Object {
            "_ref": "someAssetId",
            "_type": "reference",
          },
        },
      },
    },
    Object {
      "patch": Object {
        "id": "doc_14",
        "set": Object {
          "some.path": Object {
            "_ref": "someAssetId",
            "_type": "reference",
          },
        },
      },
    },
    Object {
      "patch": Object {
        "id": "doc_15",
        "set": Object {
          "some.path": Object {
            "_ref": "someAssetId",
            "_type": "reference",
          },
        },
      },
    },
    Object {
      "patch": Object {
        "id": "doc_16",
        "set": Object {
          "some.path": Object {
            "_ref": "someAssetId",
            "_type": "reference",
          },
        },
      },
    },
    Object {
      "patch": Object {
        "id": "doc_17",
        "set": Object {
          "some.path": Object {
            "_ref": "someAssetId",
            "_type": "reference",
          },
        },
      },
    },
    Object {
      "patch": Object {
        "id": "doc_18",
        "set": Object {
          "some.path": Object {
            "_ref": "someAssetId",
            "_type": "reference",
          },
        },
      },
    },
    Object {
      "patch": Object {
        "id": "doc_19",
        "set": Object {
          "some.path": Object {
            "_ref": "someAssetId",
            "_type": "reference",
          },
        },
      },
    },
    Object {
      "patch": Object {
        "id": "doc_20",
        "set": Object {
          "some.path": Object {
            "_ref": "someAssetId",
            "_type": "reference",
          },
        },
      },
    },
    Object {
      "patch": Object {
        "id": "doc_21",
        "set": Object {
          "some.path": Object {
            "_ref": "someAssetId",
            "_type": "reference",
          },
        },
      },
    },
    Object {
      "patch": Object {
        "id": "doc_22",
        "set": Object {
          "some.path": Object {
            "_ref": "someAssetId",
            "_type": "reference",
          },
        },
      },
    },
    Object {
      "patch": Object {
        "id": "doc_23",
        "set": Object {
          "some.path": Object {
            "_ref": "someAssetId",
            "_type": "reference",
          },
        },
      },
    },
    Object {
      "patch": Object {
        "id": "doc_24",
        "set": Object {
          "some.path": Object {
            "_ref": "someAssetId",
            "_type": "reference",
          },
        },
      },
    },
    Object {
      "patch": Object {
        "id": "doc_25",
        "set": Object {
          "some.path": Object {
            "_ref": "someAssetId",
            "_type": "reference",
          },
        },
      },
    },
    Object {
      "patch": Object {
        "id": "doc_26",
        "set": Object {
          "some.path": Object {
            "_ref": "someAssetId",
            "_type": "reference",
          },
        },
      },
    },
    Object {
      "patch": Object {
        "id": "doc_27",
        "set": Object {
          "some.path": Object {
            "_ref": "someAssetId",
            "_type": "reference",
          },
        },
      },
    },
    Object {
      "patch": Object {
        "id": "doc_28",
        "set": Object {
          "some.path": Object {
            "_ref": "someAssetId",
            "_type": "reference",
          },
        },
      },
    },
    Object {
      "patch": Object {
        "id": "doc_29",
        "set": Object {
          "some.path": Object {
            "_ref": "someAssetId",
            "_type": "reference",
          },
        },
      },
    },
    Object {
      "patch": Object {
        "id": "doc_30",
        "set": Object {
          "some.path": Object {
            "_ref": "someAssetId",
            "_type": "reference",
          },
        },
      },
    },
    Object {
      "patch": Object {
        "id": "doc_31",
        "set": Object {
          "some.path": Object {
            "_ref": "someAssetId",
            "_type": "reference",
          },
        },
      },
    },
    Object {
      "patch": Object {
        "id": "doc_32",
        "set": Object {
          "some.path": Object {
            "_ref": "someAssetId",
            "_type": "reference",
          },
        },
      },
    },
    Object {
      "patch": Object {
        "id": "doc_33",
        "set": Object {
          "some.path": Object {
            "_ref": "someAssetId",
            "_type": "reference",
          },
        },
      },
    },
    Object {
      "patch": Object {
        "id": "doc_34",
        "set": Object {
          "some.path": Object {
            "_ref": "someAssetId",
            "_type": "reference",
          },
        },
      },
    },
    Object {
      "patch": Object {
        "id": "doc_35",
        "set": Object {
          "some.path": Object {
            "_ref": "someAssetId",
            "_type": "reference",
          },
        },
      },
    },
    Object {
      "patch": Object {
        "id": "doc_36",
        "set": Object {
          "some.path": Object {
            "_ref": "someAssetId",
            "_type": "reference",
          },
        },
      },
    },
    Object {
      "patch": Object {
        "id": "doc_37",
        "set": Object {
          "some.path": Object {
            "_ref": "someAssetId",
            "_type": "reference",
          },
        },
      },
    },
    Object {
      "patch": Object {
        "id": "doc_38",
        "set": Object {
          "some.path": Object {
            "_ref": "someAssetId",
            "_type": "reference",
          },
        },
      },
    },
    Object {
      "patch": Object {
        "id": "doc_39",
        "set": Object {
          "some.path": Object {
            "_ref": "someAssetId",
            "_type": "reference",
          },
        },
      },
    },
    Object {
      "patch": Object {
        "id": "doc_40",
        "set": Object {
          "some.path": Object {
            "_ref": "someAssetId",
            "_type": "reference",
          },
        },
      },
    },
    Object {
      "patch": Object {
        "id": "doc_41",
        "set": Object {
          "some.path": Object {
            "_ref": "someAssetId",
            "_type": "reference",
          },
        },
      },
    },
    Object {
      "patch": Object {
        "id": "doc_42",
        "set": Object {
          "some.path": Object {
            "_ref": "someAssetId",
            "_type": "reference",
          },
        },
      },
    },
    Object {
      "patch": Object {
        "id": "doc_43",
        "set": Object {
          "some.path": Object {
            "_ref": "someAssetId",
            "_type": "reference",
          },
        },
      },
    },
    Object {
      "patch": Object {
        "id": "doc_44",
        "set": Object {
          "some.path": Object {
            "_ref": "someAssetId",
            "_type": "reference",
          },
        },
      },
    },
    Object {
      "patch": Object {
        "id": "doc_45",
        "set": Object {
          "some.path": Object {
            "_ref": "someAssetId",
            "_type": "reference",
          },
        },
      },
    },
    Object {
      "patch": Object {
        "id": "doc_46",
        "set": Object {
          "some.path": Object {
            "_ref": "someAssetId",
            "_type": "reference",
          },
        },
      },
    },
    Object {
      "patch": Object {
        "id": "doc_47",
        "set": Object {
          "some.path": Object {
            "_ref": "someAssetId",
            "_type": "reference",
          },
        },
      },
    },
    Object {
      "patch": Object {
        "id": "doc_48",
        "set": Object {
          "some.path": Object {
            "_ref": "someAssetId",
            "_type": "reference",
          },
        },
      },
    },
    Object {
      "patch": Object {
        "id": "doc_49",
        "set": Object {
          "some.path": Object {
            "_ref": "someAssetId",
            "_type": "reference",
          },
        },
      },
    },
    Object {
      "patch": Object {
        "id": "doc_50",
        "set": Object {
          "some.path": Object {
            "_ref": "someAssetId",
            "_type": "reference",
          },
        },
      },
    },
  ],
}
`;

exports[`batch patching (batch #2) 1`] = `
Object {
  "mutations": Array [
    Object {
      "patch": Object {
        "id": "doc_51",
        "set": Object {
          "some.path": Object {
            "_ref": "someAssetId",
            "_type": "reference",
          },
        },
      },
    },
    Object {
      "patch": Object {
        "id": "doc_52",
        "set": Object {
          "some.path": Object {
            "_ref": "someAssetId",
            "_type": "reference",
          },
        },
      },
    },
    Object {
      "patch": Object {
        "id": "doc_53",
        "set": Object {
          "some.path": Object {
            "_ref": "someAssetId",
            "_type": "reference",
          },
        },
      },
    },
    Object {
      "patch": Object {
        "id": "doc_54",
        "set": Object {
          "some.path": Object {
            "_ref": "someAssetId",
            "_type": "reference",
          },
        },
      },
    },
    Object {
      "patch": Object {
        "id": "doc_55",
        "set": Object {
          "some.path": Object {
            "_ref": "someAssetId",
            "_type": "reference",
          },
        },
      },
    },
    Object {
      "patch": Object {
        "id": "doc_56",
        "set": Object {
          "some.path": Object {
            "_ref": "someAssetId",
            "_type": "reference",
          },
        },
      },
    },
    Object {
      "patch": Object {
        "id": "doc_57",
        "set": Object {
          "some.path": Object {
            "_ref": "someAssetId",
            "_type": "reference",
          },
        },
      },
    },
    Object {
      "patch": Object {
        "id": "doc_58",
        "set": Object {
          "some.path": Object {
            "_ref": "someAssetId",
            "_type": "reference",
          },
        },
      },
    },
    Object {
      "patch": Object {
        "id": "doc_59",
        "set": Object {
          "some.path": Object {
            "_ref": "someAssetId",
            "_type": "reference",
          },
        },
      },
    },
    Object {
      "patch": Object {
        "id": "doc_60",
        "set": Object {
          "some.path": Object {
            "_ref": "someAssetId",
            "_type": "reference",
          },
        },
      },
    },
  ],
}
`;

exports[`fails if asset download fails 1`] = `
[Error: Error while fetching asset from "http://127.0.0.1:49999/img.gif":
connect ECONNREFUSED 127.0.0.1:49999]
`;

exports[`fails if asset lookup fails 1`] = `
[Error: Error while attempt to query Sanity API:
Some network err]
`;

exports[`single asset mutation 1`] = `
Object {
  "mutations": Array [
    Object {
      "patch": Object {
        "id": "movie_1",
        "set": Object {
          "metadata.poster": Object {
            "_ref": "someAssetId",
            "_type": "reference",
          },
        },
      },
    },
  ],
}
`;

exports[`single create mutation 1`] = `
Object {
  "mutations": Array [
    Object {
      "patch": Object {
        "id": "movie_1",
        "set": Object {
          "metadata.poster": Object {
            "_ref": "newAssetId",
            "_type": "reference",
          },
        },
      },
    },
  ],
}
`;
